pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +449,705 @@ pub mod pallet {
+			self.reset_top_data::<T>(&top_delegations);
+			self.delegation_count -= 1u32;
+			<TopDelegations<T>>::insert(candidate, top_delegations);
+			// return whether total counted changed
+			Ok(old_total_counted == self.total_counted)
+		}
+		/// Remove bottom delegation
+		/// Returns if_total_counted_changed: bool
+		pub fn rm_bottom_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			// remove bottom delegation
+			let mut bottom_delegations =
+				<BottomDelegations<T>>::get(candidate).expect("existence proof TODO");
+			let mut actual_amount_option: Option<BalanceOf<T>> = None;
+			bottom_delegations.delegations = bottom_delegations
+				.delegations
 				.clone()
 				.into_iter()
 				.filter_map(|d| {
 					if d.owner != delegator {
 						Some(d)
 					} else {
-						delegation_amt = Some(d.amount);
+						actual_amount_option = Some(d.amount);
 						None
 					}
 				})
 				.collect();
-			// if err, no item with account exists in top || bottom
-			let amount = delegation_amt.ok_or(Error::<T>::DelegatorDNEinTopNorBottom)?;
-			self.total_backing -= amount;
-			Ok((false, amount))
-		}
-		/// Return true if in_top after call
-		/// Caller must verify before call that account is a delegator
-		fn increase_delegation(&mut self, delegator: A, more: B) -> bool {
-			let mut in_top = false;
-			for x in &mut self.top_delegations {
-				if x.owner == delegator {
-					x.amount += more;
-					self.total_counted += more;
-					self.total_backing += more;
-					in_top = true;
-					break;
+			let actual_amount = actual_amount_option.ok_or(Error::<T>::DelegationDNE)?;
+			bottom_delegations.total -= actual_amount;
+			// update candidate info
+			self.reset_bottom_data::<T>(&bottom_delegations);
+			self.delegation_count -= 1u32;
+			<BottomDelegations<T>>::insert(candidate, bottom_delegations);
+			Ok(false)
+		}
+		/// Increase delegation amount
+		pub fn increase_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+			bond: BalanceOf<T>,
+			more: BalanceOf<T>,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let lowest_top_eq_highest_bottom =
+				self.lowest_top_delegation_amount == self.highest_bottom_delegation_amount;
+			let bond_geq_lowest_top = bond.into() >= self.lowest_top_delegation_amount;
+			let delegation_dne_err: DispatchError = Error::<T>::DelegationDNE.into();
+			if bond_geq_lowest_top && !lowest_top_eq_highest_bottom {
+				// definitely in top
+				self.increase_top_delegation::<T>(candidate, delegator.clone(), more)
+			} else if bond_geq_lowest_top && lowest_top_eq_highest_bottom {
+				// update top but if error then update bottom (because could be in bottom because
+				// lowest_top_eq_highest_bottom)
+				let result = self.increase_top_delegation::<T>(candidate, delegator.clone(), more);
```

Comment:
Breadcrumb for me: make sure this can only error in the way expected

Reply:
Update: yes, it does, but it might be prudent to ensure the error is the proper "DNE"  variant